### PR TITLE
doc: Document the endpoint lifecycle and reserved:init identity

### PR DESCRIPTION
**Pull request by rlenglet**
Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4212)

### CODE REVIEW

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-me-please

File: Documentation/policy/lifecycle.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+Every endpoint in Cilium is in one of the following states:
+
+* ``restoring``: The endpoint was started before Cilium started, and
+  Cilium is restoring its networking configuration.
+* ``waiting-for-identity``: Cilium is allocating a unique identity for
+  the endpoint.
+* ``waiting-to-regenerate``: The endpoint received an identity and is
+  waiting for its networking configuration to be (re)generated.
+* ``regenerating``: The endpoint's networking configuration is being
+  (re)generated. This includes programming PBF for that endpoint.
```
Comment (joestringer) [inline Documentation/policy/lifecycle.rst]: `s/PBF/BPF/`

**Review by joestringer — COMMENTED** [review]
Mostly did a spellcheck, description seems right to me.

Question for the picture - can't `restoring` transition into `waiting-for-identity`? What if all other endpoints with the same labels are removed during a restart? I guess the identity is still leased in the kvstore, so the identity should be allocated in the kvstore, then Cilium can almost immediately fetch the identity... except what if there is also a kvstore connectivity issue at the same time? Will we restore and handle traffic assuming that the previous numeric identity is the correct representation of the identity for the endpoint's labels?

**Comment by rlenglet** [comment]
> Question for the picture - can't restoring transition into waiting-for-identity?

Good question. I just drew what the code specifies. I don’t know why that transition is not allowed or not required.

**Comment by rlenglet** [comment]
test-docs-please

File: Documentation/policy/lifecycle.rst (hunk 1 above)
Comment (rlenglet) [reply Documentation/policy/lifecycle.rst]: Fixed.

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
test-me-please

File: Documentation/policy/lifecycle.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+The state of an endpoint can be queried using the ``cilium endpoint
+list`` and ``cilium endpoint get`` CLI commands.
+
+While an endpoint is running, it transitions between the
+``waiting-for-identity``, ``waiting-to-regenerate``, ``regenerating``,
+and ``ready`` states.  A transition into the ``waiting-for-identity``
+state indicates that the endpoint changed its identity.  A transition
+into the ``waiting-to-regenerate`` or ``regenerating`` state indicates
+that the policy to be enforced on the endpoint has changed because of
+an identity change or a policy change.
```
Comment (ianvernon) [inline Documentation/policy/lifecycle.rst]: I'd recommend mentioning configuration changes (i.e., Conntrack configuration, endpoint-specific configuration) that translate into changes to `#define` in the headerfiles for specific options also trigger regeneration.

**Review by ianvernon — COMMENTED** [review]
Thanks for taking care of the documentation of the endpoint state machine - we have needed this for some time!

**Review by aanm — APPROVED** [review]
Nice!

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-docs-please

**Comment by rlenglet** [comment]
Addressed review comments.

**Comment by rlenglet** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**